File: 110-balanced-binary-tree/balanced-binary-tree.cpp
// //O(n)
// class Solution {
// public:
//     bool isBalanced(TreeNode *root){
//         return dfsHt(root)!=-1;
//     }
//     int dfsHt(TreeNode *root){
//         if(root==NULL) return 0;
        
//         int lht=dfsHt(root->left);
//         if(lht==-1) return -1;
//         int rht=dfsHt(root->right);
//         if(rht==-1) return -1;
        
//         if(abs(lht-rht)>1) return -1;
//         return max(lht,rht)+1;
//     }
// };
class Solution {
public:
    int ht(TreeNode *root){
        if(!root) return 0;
        return max(ht(root->left),ht(root->right))+1;
    }
    bool isBalanced(TreeNode *root){
        if(!root) return true;
        int left=ht(root->left);
        int right=ht(root->right);
        return abs(left-right) <=1 && isBalanced(root->left) && isBalanced(root->right);
    }
}; 

File: 110-balanced-binary-tree/balanced-binary-tree.py
class Solution(object):
    def isBalanced(self, root):
        def dfs(root):
            if(not root):
                return [True,0]
            left,right=dfs(root.left),dfs(root.right)
            balanced=left[0] and right[0] and abs(left[1]-right[1]) <=1
            return [balanced,1+max(left[1],right[1])]
        return dfs(root)[0]

# class Solution(object):
#     def isBalanced(self, root):
#         return (self.Height(root) >= 0)
#     def Height(self, root):
#         if root is None:  
#             return 0
#         leftheight, rightheight = self.Height(root.left), self.Height(root.right)
#         if leftheight < 0 or rightheight < 0 or abs(leftheight - rightheight) > 1:
#             return -1
#         return max(leftheight, rightheight) + 1